Sprint 2/step 2

pavelborisofff opened the pull request:
Добрый день!

Местами не справился с типизацией в ТС, но так по большей части похоже на правду, надеюсь :) Код далёк от совершенства, конечно…

Жду  комментариев :)

pavelborisofff committed:
small fixes and reformat code

pavelborisofff committed:
use redux to store states

pavelborisofff committed:
Merge pull request #5 from pavelborisofff/probe

use redux to store states

gennady-bars commented on src/services/index.ts:
```diff
@@ -0,0 +1,16 @@
+});
+
+export type RootState = ReturnType<typeof store.getState>
+export type AppDispatch = typeof store.dispatch
```
Отлично типизировали стейт, диспатч

gennady-bars commented on src/services/reducers/recipeReducer.ts:
```diff
@@ -0,0 +1,84 @@
+const recipeReducer = (state = initialRecipe, action: any):IRecipeState => {
+  switch (action.type) {
+    case INGREDIENT_ADD:
+      action.payload.uuid = uuidv4();
```
`      action.payload.uuid = uuidv4();
`

Редьюсер должен быть **чистой функцией**.  Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста.

gennady-bars commented on src/services/actions/ingredientsActions.ts:
```diff
@@ -0,0 +1,58 @@
+}
+
+
+export const getIngredients = () => async (dispatchFunc: ({type}:IIngredients) => void) => {
```
`dispatchFunc: ({type}:IIngredients) => void)`


у Вас уже есть тип для диспатча


```js

export const getIngredients = () => async (dispatchFunc: AppDispatch) => {


```


и обычно этот аргумент просто называют `dispatch`


```js

export const getIngredients = () => async (dispatch: AppDispatch) => {


```

pavelborisofff replied:
Хм… ага, я этот вариант типизации брал из документации Redux Toolkit но не применил его где надо

gennady-bars commented on src/components/main/main.tsx:
```diff
@@ -1,68 +1,50 @@
+          </p>
+        )}
+        {!isLoading && !isError &&
+          <>
```
`<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку

pavelborisofff replied:
Исправил, пропустил в прошлый раз

gennady-bars commented on src/components/modal/modal.tsx:
```diff
@@ -14,8 +17,8 @@ type ModalProps = {
+const modalRoot = document.getElementById("modal-root") as HTMLElement;
+const body = document.querySelector("body") as HTMLElement;
 
 const Modal = ({ children, modalHeading, showModal, onClose }: ModalProps) => {
```
У Вас не должно быть 2х компонентов `Modal`. Нужно удалить теперь лишний


Через редакс сделали достаточно универсально, поэтому можно оставить только `ModalRedux`, назвав его просто `Modal`

gennady-bars commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -1,169 +1,100 @@
+    const body = {
+      ingredients: [...usedIngredients.map(item => item._id), bun?._id || null]
+    }
+    dispatch(orderPost(body as any) as any); // TODO: не знаю, что тут сделать, чтобы не было ошибки в TS
```
**Можно лучше**

Это потом уже изучите


Но для этого нужно типизировать редакс и `thunk`



вот ссылка на документацию по его типизации. https://redux.js.org/usage/usage-with-typescript#type-checking-redux-thunks

gennady-bars commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -1,169 +1,100 @@
+        > 
+          {!usedIngredients && <p className={`${styles.empty} text text_type_main-medium ml-8 p-10`}>Перетащите ингридиенты сюда</p>}
+          {usedIngredients && usedIngredients.map((item, index) => (
+            <div className={`${styles.constructorItem}`} key={item.uuid}>
```
`key={item.uuid}`


Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

gennady-bars commented on src/components/burger-constructor/recipe-item.jsx:
```diff
@@ -0,0 +1,83 @@
+  dragItem(dropItem(itemRef));
+
+  return (
+    <div className={`${styles.constructorItem}`} key={item.uuid} ref={itemRef}>
```
`key={item.uuid}`

Атрибут `key` нужен только при отрисовке **массивов** компонентов. Просто так внутри компонентов не нужно его указывать.

gennady-bars commented on src/services/actions/ingredientsActions.ts:
```diff
@@ -0,0 +1,58 @@
+    const response:IIngredientsResponse = await request({endpoint: API.ingredients});
+    const preparedData:IPreparedData = sortByTypes(response);
+    dispatchFunc({type: INGREDIENTS_REQUEST_SUCCESS,ingredients: preparedData});
+  } catch (error) {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Отлично типизировали стейт, диспатч 
* Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`
* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Хорошая логика редакса и `DnD`

но есть некоторые недочеты: 


* По заданию нужно сделать логику **переключения табов**, когда идет **скролл** ингредиентов в списке слева. Это сделать можно с помощью библиотеки `react-intersection-observer`
* Редьюсер должен быть **чистой функцией**.  Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста.
* `<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку
* Атрибут `key` нужен только при отрисовке **массивов** компонентов. Просто так внутри компонентов не нужно его указывать.
* 



**Можно лучше**

* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
*  желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа
* Чтобы **не дублировать** типы экшенов и сами объекты с `type`, принято выносить их в отдельные функции, называемые **`action creator`** (функция, которая возвращает объект с `type`). Это позволяет сделать **понятное название** для функции, которое будет говорить, что делает она, и избавит от повтора объекта с `type` в коде. Подробнее тут: [https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#генераторы-экшенов-action-creators](https://rajdee.gitbooks.io/redux-in-russian/content/docs/basics/Actions.html#%D0%B3%D0%B5%D0%BD%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D1%8D%D0%BA%D1%88%D0%B5%D0%BD%D0%BE%D0%B2-action-creators)



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

pavelborisofff committed:
make active tabs on scroll

pavelborisofff committed:
1. blocking order button if not bun in recipe
2. clear recipe if order has created
3. clear order data when closing modal with order's info
4. small fixes and refactor

pavelborisofff committed:
Merge pull request #7 from pavelborisofff/probe

Probe

pavelborisofff commented:
Геннадий, добрый день!

Я в ветке probe работал и смерджил её с текущим Sprint-2/step-2
не знаю, насколько это правильным решением было…

Про вашим комментам  недочёты убрал, по "можно лучше" тоже прошёлся — но только c экшн криеэйторами не разобрался пока до конца.

gennady-bars commented on src/components/modal/modal.tsx:
```diff
@@ -9,63 +9,22 @@ import styles from "./modal.module.scss";
   
-  const onClose = () => dispatch({ type: MODAL_CLOSE });
+  const onClose = () => {
+    if (orderNumber) dispatch({ type: ORDER_RESET });
```
`    if (orderNumber) dispatch({ type: ORDER_RESET });
`

это нужно удалить 


`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.


очищение заказа нужно делать в блоке `then` после получения номера заказа, а не при закрытии

pavelborisofff replied:
Убрал отсюда состояние заказа, перенёс ORDER_RESET в burger-constructor.tsx перед созданием нового.

gennady-bars commented:
> Я в ветке probe работал и смерджил её с текущим Sprint-2/step-2


ну, по заданию нужно делать ветку под спринт. Внизу задания всегда есть название ветки, которую нужно сделать для ревью. А потом (после полного принятия работы спринта) сливать ее в `main`

gennady-bars reviewed (commented):
Уже почти все отлично,  но:

* вот это можно было исправить https://github.com/pavelborisofff/react-burger/pull/6#discussion_r1118021318
* `Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.
* и это не исправили https://github.com/pavelborisofff/react-burger/pull/6#discussion_r1118021451

pavelborisofff commented on src/components/modal/modal.tsx:
```diff
@@ -1,60 +1,64 @@
+  
+  const onClose = () => {
+    if (orderNumber) dispatch({ type: ORDER_RESET });
+    dispatch({ type: MODAL_CLOSE })
```
Убрал отсюда состояние заказа, перенёс ORDER_RESET в burger-constructor.tsx перед созданием нового.

pavelborisofff committed:
1. remove component state from modal
2. fix dispatch type
3. remove unused react fragments

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
